File: Tutorials/ASP.NET/MVC5/CS/Controllers/OrderController.cs
using System.Linq;
using System.Web.Mvc;
using DevExpress.Xpo;
using AspNetMvcApplication.Models;
using XpoTutorial;

namespace AspNetMvcApplication.Controllers {

    public class OrderController : BaseController {

        [ValidateInput(false)]
        public ActionResult OrderListPartial(int customerOid) {
            var model = UnitOfWork.Query<Customer>()
                .Select(t => new CustomerViewModel() {
                    Oid = t.Oid,
                    FirstName = t.FirstName,
                    LastName = t.LastName,
                    Orders = t.Orders.Select(o => new OrderViewModel() {
                        Oid = o.Oid,
                        OrderDate = o.OrderDate,
                        ProductName = o.ProductName,
                        Freight = o.Freight,
                        CustomerId = o.Customer.Oid
                    }).ToList()
                }).FirstOrDefault(t => t.Oid == customerOid);
            ViewData["CustomersList"] = UnitOfWork.Query<Customer>()
                .Select(t => new CustomerViewModel() {
                    Oid = t.Oid,
                    FirstName = t.FirstName,
                    LastName = t.LastName
                }).OrderBy(t => t.FirstName).ThenBy(t => t.LastName).ToList();
            return PartialView("OrderListPartial", model);
        }


        [HttpPost, ValidateInput(false)]
        public ActionResult AddOrder(int customerOid, OrderViewModel model) {
            if(ModelState.IsValid) {
                SafeExecute(() => {
                    var order = new Order(UnitOfWork) {
                        OrderDate = model.OrderDate,
                        ProductName = model.ProductName,
                        Freight = model.Freight,
                        Customer = UnitOfWork.GetObjectByKey<Customer>(model.CustomerId)
                    };
                    UnitOfWork.CommitChanges();
                });
            } else {
                ViewData["EditError"] = "Please, correct all errors.";
            }
            return OrderListPartial(customerOid);
        }

        [HttpPost, ValidateInput(false)]
        public ActionResult UpdateOrder(int customerOid, OrderViewModel model) {
            if(ModelState.IsValid) {
                SafeExecute(() => {
                    var order = UnitOfWork.GetObjectByKey<Order>(model.Oid);
                    order.OrderDate = model.OrderDate;
                    order.ProductName = model.ProductName;
                    order.Freight = model.Freight;
                    order.Customer = UnitOfWork.GetObjectByKey<Customer>(model.CustomerId);
                    UnitOfWork.CommitChanges();
                });
            } else {
                ViewData["EditError"] = "Please, correct all errors.";
            }
            return OrderListPartial(customerOid);
        }

        [HttpPost, ValidateInput(false)]
        public ActionResult DeleteOrder(int customerOid, int Oid) {
            SafeExecute(() => {
                var order = UnitOfWork.GetObjectByKey<Order>(Oid);
                order.Delete();
                UnitOfWork.CommitChanges();
            });
            return OrderListPartial(customerOid);
        }
    }
}

File: Tutorials/ASP.NET/MVC5/VB/Controllers/OrderController.vb
Imports System.Linq
Imports DevExpress.Xpo
Imports AspNetMvcApplication.Models
Imports XpoTutorial
Imports System.Web.Mvc

Namespace AspNetMvcApplication.Controllers

    Public Class OrderController
        Inherits BaseController

        <ValidateInput(False)>
        Public Function OrderListPartial(ByVal customerOid As Integer) As ActionResult
            Dim model = UnitOfWork.Query(Of Customer)().Select(Function(t) New CustomerViewModel() With {
                .Oid = t.Oid,
                .FirstName = t.FirstName,
                .LastName = t.LastName,
                .Orders = t.Orders.Select(Function(o) New OrderViewModel() With {
                    .Oid = o.Oid,
                    .OrderDate = o.OrderDate,
                    .ProductName = o.ProductName,
                    .Freight = o.Freight,
                    .CustomerId = o.Customer.Oid
                }).ToList()
            }).FirstOrDefault(Function(t) t.Oid = customerOid)
            ViewData("CustomersList") = UnitOfWork.Query(Of Customer)().Select(Function(t) New CustomerViewModel() With {
                .Oid = t.Oid,
                .FirstName = t.FirstName,
                .LastName = t.LastName
            }).OrderBy(Function(t) t.FirstName).ThenBy(Function(t) t.LastName).ToList()
            Return PartialView("OrderListPartial", model)
        End Function


        <HttpPost, ValidateInput(False)>
        Public Function AddOrder(ByVal customerOid As Integer, ByVal model As OrderViewModel) As ActionResult
            If ModelState.IsValid Then
                SafeExecute(Sub()
                                Dim order = New Order(UnitOfWork) With {
                                    .OrderDate = model.OrderDate,
                                    .ProductName = model.ProductName,
                                    .Freight = model.Freight,
                                    .Customer = UnitOfWork.GetObjectByKey(Of Customer)(model.CustomerId)
                                }
                                UnitOfWork.CommitChanges()
                            End Sub)
            Else
                ViewData("EditError") = "Please, correct all errors."
            End If
            Return OrderListPartial(customerOid)
        End Function

        <HttpPost, ValidateInput(False)>
        Public Function UpdateOrder(ByVal customerOid As Integer, ByVal model As OrderViewModel) As ActionResult
            If ModelState.IsValid Then
                SafeExecute(Sub()
                                Dim order = UnitOfWork.GetObjectByKey(Of Order)(model.Oid)
                                order.OrderDate = model.OrderDate
                                order.ProductName = model.ProductName
                                order.Freight = model.Freight
                                order.Customer = UnitOfWork.GetObjectByKey(Of Customer)(model.CustomerId)
                                UnitOfWork.CommitChanges()
                            End Sub)
            Else
                ViewData("EditError") = "Please, correct all errors."
            End If
            Return OrderListPartial(customerOid)
        End Function

        <HttpPost, ValidateInput(False)>
        Public Function DeleteOrder(ByVal customerOid As Integer, ByVal Oid As Integer) As ActionResult
            SafeExecute(Sub()
                            Dim order = UnitOfWork.GetObjectByKey(Of Order)(Oid)
                            order.Delete()
                            UnitOfWork.CommitChanges()
                        End Sub)
            Return OrderListPartial(customerOid)
        End Function
    End Class
End Namespace